### PR TITLE
Fix for #4136 connection to linux refused by using wrong Xauthority file

### PR DESCRIPTION
**Pull request by awalon**
Solution:
1. Get file from sddm-greeter and startplasma-x11 (aka KDE) processes as well
2. Get users home dir from system instead of hard coded pattern /home/<user>
3. Prefer XAUTHORITY environment variable over Xorg -auth if available

### CODE REVIEW

File: libs/hbb_common/src/config.rs [hunk 1]
```diff
@@ -322,7 +322,19 @@ fn patch(path: PathBuf) -> PathBuf {
                         .trim()
                         .to_owned();
                     if user != "root" {
-                        return format!("/home/{user}").into();
+                        let cmd = format!("getent passwd '{}' | awk -F':' '{{print $6}}'", user);
+                        if let Ok(output) = std::process::Command::new(cmd).output() {
```
Comment (rustdesk) [inline libs/hbb_common/src/config.rs:326]: How about if here is not OK?

File: src/platform/linux.rs [hunk 2]
```diff
@@ -449,12 +451,37 @@ pub fn get_active_username() -> String {
     get_values_of_seat0(&[2])[0].clone()
 }
 
+pub fn get_user_home_by_name(username: &str) -> Option<PathBuf> {
+    return match get_user_by_name(username) {
+        None => {
+            log::debug!("no uid found for user: {}", username);
+            None }
```
Comment (rustdesk) [inline src/platform/linux.rs]: format

File: src/platform/linux.rs [hunk 3]
```diff
@@ -824,56 +862,107 @@ mod desktop {
...
-                        } else if auth_found {
-                            if std::path::Path::new(v).is_absolute() {
-                                self.xauth = v.to_string();
-                            } else {
-                                if let Some(pid) = line.split_whitespace().nth(1) {
-                                    let home_dir = get_env_from_pid("HOME", pid);
-                                    if home_dir.is_empty() {
-                                        self.xauth = format!("/home/{}/{}", self.username, v);
+
+                    if let Some(pid) = line.split_whitespace().nth(1) {
```
Comment (rustdesk) [inline src/platform/linux.rs]: Why add one more `if` here?

File: src/platform/linux.rs [hunk 4]
```diff
@@ -824,56 +862,107 @@ mod desktop {
...
+                                        } else {
+                                            base_dir = home_dir;
+                                        }
+                                        if Path::new(&base_dir).exists() {
+                                            self.xauth = format!("{}/{}", base_dir, v);
+                                            log::debug!("+ auth file with home path added: {}", self.xauth);
+                                        };
                                     } else {
-                                        self.xauth = format!("{}/{}", home_dir, v);
+                                        // unreachable!
```
Comment (rustdesk) [inline src/platform/linux.rs]: Can you explain what changes here line870 -> line 895?

File: libs/hbb_common/src/config.rs (hunk 1 above)
Comment (awalon) [reply libs/hbb_common/src/config.rs:326]: Good point missed that case, will be fixed by followup PR

File: src/platform/linux.rs (hunk 2 above)
Comment (awalon) [reply src/platform/linux.rs]: Missing line breake (after 2nd None) by updated PR as well

File: src/platform/linux.rs (hunk 3 above)
Comment (awalon) [reply src/platform/linux.rs]: Thanks for review, that's a left over for Xauth file fallback. Cann t test this here so this step will be  removed again with next PR.

File: src/platform/linux.rs (hunk 4 above)
Comment (awalon) [reply src/platform/linux.rs]: Highlevel: 
As mentioned by newly added debug messages auth- file will be extraxcted from Xorg process and for a relative path home path will be added.

If no HOME var was found the home folder for given user will be determined from system config (new function: get_user_home_by_name)  old logic just used hard coded  default pattern "/home/<username>" (now only if no system config was found)  which wasn't working for domain users (AD, samba, freeipa, ...) where an additional domain subfolder (/home/<domain>/<username>) needs to be added.

In addition detected file needs to exist, if not this loop won't detect any auth- file. So next step can detect a valid file.

Next step (Todo): Detect auth file by checking process tree (processes below Xorg) without need of predefined checks for Gname, XFCE, Plasma, ...). But as this was my first modifiaction using rust language, there was too less time for better solution.

File: libs/hbb_common/src/config.rs [hunk 5]
```diff
@@ -322,6 +322,17 @@ fn patch(path: PathBuf) -> PathBuf {
...
                         .to_owned();
                     if user != "root" {
+                        let cmd = format!("getent passwd '{}' | awk -F':' '{{print $6}}'", user);
+                        if let Ok(output) = std::process::Command::new(cmd).output() {
+                            let home_dir = String::from_utf8_lossy(&output.stdout)
+                                .to_string()
+                                .trim()
+                                .to_owned();
+                            if !home_dir.is_empty() {
+                                log::info!("config::patch: got home dir from: {}", home_dir);
```
Comment (rustdesk) [inline libs/hbb_common/src/config.rs]: log::debug or remove please.

File: src/platform/linux.rs [hunk 6]
```diff
@@ -824,22 +863,36 @@ mod desktop {
                 "ps -u {} -f | grep 'Xorg' | grep -v 'grep'",
                 &self.uid
             )) {
+                log::debug!("extracting xauth file for Xorg running with uid: {}", self.uid);
```
Comment (rustdesk) [inline src/platform/linux.rs]: Too many log::debug! which it will be print to file, We'd better remove these log::debug before code merge.

**Review by rustdesk — CHANGES_REQUESTED** [review]

File: src/platform/linux.rs (hunk 6 above)
Comment (awalon) [reply src/platform/linux.rs]: Debug messages was removed now

File: libs/hbb_common/src/config.rs (hunk 5 above)
Comment (awalon) [reply libs/hbb_common/src/config.rs]: log message removed

**Comment by rustdesk** [comment]
#4136

**Comment by rustdesk** [comment]
I guess https://github.com/rustdesk/rustdesk/issues/440 has the same, no idea what process it is.

**Comment by rustdesk** [comment]
#440

**Pull request merged**